### ISSUE TITLE
Properly send handshake error to client

### ISSUE THREAD

**Issue by fourpixels** [socketio/socket.io#2499]
Currently, if the handshake fails, the code that's get executed is this:

``` js
if (err) return socket.error(err.data || err.message);
```

[https://github.com/socketio/socket.io/blob/master/lib/namespace.js#L160](https://github.com/socketio/socket.io/blob/master/lib/namespace.js#L160)

So on the client side, the only thing reaching most of the time is plain string (as common Errors do not have a `data` property).

And because I would like to catch that error and handle it properly, I **must** compare strings, or embed JSON object inside, which is total nonsense.

**Comment by didaxRedux** [comment]
I agree with @fourpixels 
Probably the best choice is to pass an object with "message" and "code" properties.
"code" can be used on client side to handle error in a better way.

**Comment by FoghostCn** [comment]
I'd like to see this change ,currently I have to embed an object to message property

**Comment by fourpixels** [comment]
What I've done so far is using my own custom Error class (commonly inherited from the regular Error) and I've just hacked a `data` property. On creation, I automatically (internally) set the data property to some kind of serialized data (message, code, whatever I've got).
It started working automatically, but in my case I already got my own custom Error class, so I just added the new property. If you haven't, then JSON serialization seems easiest solution to me.

I'm a bit afraid that having in mind pace things get resolved around should mean this won't get fixed soon.
